libbeat/outputs/fileout/file.go

```diff
@@ -76,16 +80,30 @@ func (out *fileOutput) PublishEvent(
 	opts outputs.Options,
 	data outputs.Data,
 ) error {
-	jsonEvent, err := json.Marshal(data.Event)
-	if err != nil {
-		// mark as success so event is not sent again.
-		op.SigCompleted(sig)
+	var serializedEvent []byte
+	var err error
+
+	if out.format != nil {
+		formattedEvent, err := out.format.Run(data.Event)
+		if err != nil {
+			logp.Err("Fail to format event (%v): %#v", err, data.Event)
+			op.SigCompleted(sig)
+			return err
+		}
+		serializedEvent = []byte(formattedEvent)
 
-		logp.Err("Fail to json encode event(%v): %#v", err, data.Event)
-		return err
+	} else {
+		serializedEvent, err = json.Marshal(data.Event)
+		if err != nil {
+			// mark as success so event is not sent again.
+			op.SigCompleted(sig)
+
+			logp.Err("Fail to json encode event(%v): %#v", err, data.Event)
+			return err
+		}
```

Comment:
This kind of encoding/formatting strategy seems to become a common pattern. Consider introducing some abstraction via of 'encoder'/'formatter' interface/function instead of  adding if/else branches. Should get us a more clean error handling again.